apiextensions: refactor crd conversion e2e into apiextensions integration test

sttts opened the pull request:
Based on https://github.com/kubernetes/kubernetes/pull/77756, this adds the e2e CRD conversion tests to the apiextensions-apiserver integration tests. This is the preparation for https://github.com/kubernetes/kubernetes/pull/77743 and https://github.com/kubernetes/kubernetes/pull/77922.

sttts commented:
/assign @liggitt @jpbetz

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/conversion_test.go:
```diff
@@ -0,0 +1,274 @@
+	}
+
+	// After changing a CRD, the resources for versions will be re-created that can be result in
+	// cancelled connection (e.g. "grpc connection closed" or "context canceled").
```
this shouldn't be true after https://github.com/kubernetes/kubernetes/pull/77816

sttts replied:
will leave it until #77816 is merged. Otherwise this flakes.

liggitt replied:
It's already merged

sttts replied:
It merged. Removed this workaround.

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/conversion_test.go:
```diff
@@ -0,0 +1,274 @@
+		updateCRD.Spec.Versions[0].Storage = false
+		updateCRD.Spec.Versions[1].Storage = true
+
+		updateCRD, err = apiextensionsClient.ApiextensionsV1beta1().CustomResourceDefinitions().Update(updateCRD)
```
do we need to wait for the handler to update?

sttts replied:
yes. How can we without trying again and again while checking etcd or by checking via the webhook?

sttts replied:
Added the TODO at https://github.com/kubernetes/kubernetes/pull/77868/files#diff-7dafb2a7b5eb7c3edad4fdc93db9a243R179. https://github.com/kubernetes/kubernetes/pull/77756/files#diff-7dafb2a7b5eb7c3edad4fdc93db9a243R350 brings a variant with etcd access code. We should use that.

sttts replied:
fixed by rebasing on @jpbetz's test framework.

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/conversion_test.go:
```diff
@@ -0,0 +1,274 @@
+	name2 := "cr-instance-2"
+	t.Log("Creating a v1 custom resource")
+
+	// cr-instance-2 is stored as v1. Now change storage version to v2.
```
is this comment accurate? I only see us creating "cr-instance-2" after switching storage version to v2

sttts replied:
typo

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/conversion_test.go:
```diff
@@ -0,0 +1,274 @@
+	// cr-instance-2 is stored as v1. Now change storage version to v2.
+	var updateCRD *apiextensionsv1beta1.CustomResourceDefinition
+	var err error
+	for i := 0; i < 5; i++ {
```
use RetryOnConflict?

sttts replied:
done

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/conversion_test.go:
```diff
@@ -0,0 +1,274 @@
+	}
+	assert.Equal(t, 2, len(list.Items))
+	assert.True(t,
+		(list.Items[0].GetName() == name1 && list.Items[1].GetName() == name2) ||
```
testCRListConversion is depending on "cr-instance-1" already having been created in v1?

sttts replied:
yes

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/conversion_test.go:
```diff
@@ -0,0 +1,274 @@
+	url := webhookServer.URL + "/crdconvert"
+	crd := multiVersionConversionFixture.DeepCopy()
+	crd.Spec.Conversion = &apiextensionsv1beta1.CustomResourceConversion{
+		Strategy: apiextensionsv1beta1.WebhookConverter,
```
do you want to add a structura schema here so you don't have to when that becomes a requirement for webhookconversion?

sttts replied:
done

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/converter/BUILD:
```diff
@@ -0,0 +1,50 @@
+)
+
+go_test(
+    name = "go_default_test",
```
need to add `tags = ["integration"],` for this to show up in the right CI test suite

liggitt replied:
(or possibly `tags = ["integration","etcd"],` if it also depends on external etcd?)

jpbetz commented:
I also opened https://github.com/kubernetes/kubernetes/pull/77756 last night. The test cases are entirely different, but it also adds a file with the exact `conversion_test.go` name.

sttts commented on test/e2e/apimachinery/crd_conversion_webhook.go:
```diff
@@ -374,6 +375,8 @@ func testCRListConversion(f *framework.Framework, testCrd *crd.TestCrd) {
 	// cancelled connection (e.g. "grpc connection closed" or "context canceled").
 	// Just retrying fixes that.
+	//
+	// TODO: we have to wait for the storage version to become effective. Storage version changes are not instant.
```
here is the same issue with the storage change race.

liggitt commented:
/retest

integration flake fixed in https://github.com/kubernetes/kubernetes/pull/77877

liggitt commented:
/retest

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/conversion_test.go:
```diff
@@ -0,0 +1,281 @@
+		},
+	}
+
+	// TODO: we have to wait for the storage version to become effective. Storage version changes are not instant.
```
if we know we have a race, we should at least add a `time.Sleep(time.Second)` here until we get a more complete way of dealing with it. I'm not eager to add a flake to our integration suite.

sttts replied:
added a sleep. But why should that flake? I believe it will run through, but will "just" not test what it should.

roycaihw replied:
just to give some data point, the old e2e test didn't look flaky from https://testgrid.k8s.io/google-gce#gci-gce-alpha-features, but maybe it's because e2e test is slower

roycaihw commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/conversion_test.go:
```diff
@@ -0,0 +1,284 @@
+/*
+Copyright 2018 The Kubernetes Authors.
```
2019

sttts replied:
fixed

roycaihw commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/conversion_test.go:
```diff
@@ -0,0 +1,284 @@
+		versionedClients[v.Name] = dynamicClient.Resource(schema.GroupVersionResource{Group: crd.Spec.Group, Version: v.Name, Resource: crd.Spec.Names.Plural})
+	}
+	testCustomResourceConversionWebhook(t, crd, versionedClients)
+	testCRListConversion(t, crd, apiExtensionClient, versionedClients)
```
nit: clean up the cluster scoped CRD and CRs? 

we are using a unique etcd-prefix so maybe it's fine

sttts replied:
we can do all that in a follow. @jpbetz has his toughts about a good test structure probably. These are only the tests from e2e to make them available here and not to regress. Would leave it for the moment.

sttts replied:
rebased on @jpbetz's test framework.

roycaihw reviewed (commented):
integration test framework doesn't play well with unit test in the path (see bottom of [200404.stdout](https://gcsweb.k8s.io/gcs/kubernetes-jenkins/pr-logs/pull/77868/pull-kubernetes-integration/1128386352505688065/artifacts/)):
```
ok      k8s.io/kubernetes/vendor/k8s.io/apiextensions-apiserver/test/integration        202.785s
flag provided but not defined: -alsologtostderr
Usage of /tmp/go-build527842902/b2250/converter.test:
  -test.bench regexp
...
FAIL    k8s.io/kubernetes/vendor/k8s.io/apiextensions-apiserver/test/integration/converter      0.106s
```

yue9944882 commented:
/retest

sttts commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/helpers.go:
```diff
@@ -95,6 +97,78 @@ func UpdateCustomResourceDefinitionWithRetry(client clientset.Interface, name st
+	if err != nil {
+		t.Fatalf("Failed to create %s CRD; %v", crd.Name, err)
+	}
+	crd = createdCRD
```
true. Before the `testing.T` error handling it made sense.

roycaihw commented:
it's still failing the [unit test dir](https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/77868/pull-kubernetes-integration/1128615323457556484/)

```
# local reproduce
$ make test-integration WHAT=./staging/src/k8s.io/apiextensions-apiserver/test/integration/converter
flag provided but not defined: -alsologtostderr
...
FAIL	k8s.io/kubernetes/staging/src/k8s.io/apiextensions-apiserver/test/integration/converter	0.078s
```

maybe you could move the convert package somewhere else, or whitelist it in https://github.com/kubernetes/kubernetes/blob/master/hack/make-rules/test-integration.sh#L47

sttts committed:
apiextensions: add storage race TODO in CRD conversion e2e test

sttts committed:
apiextensions: make integration test server flags available to test code

sttts committed:
apiextensions: fix non-trivial conversion integration test

This adds a third version v1alpha1 which has the same schema as v1beta1. Moreover, v1beta1 becomes the storage version. Hence, we can do noop webhook conversion from v1alpha1 to v1beta1 and back.

sttts commented:
/retest

sttts commented:
Rebased on @jpbetz's test framework.

@jpbetz please look especially at the 2nd and 3rd last commits. The non-trivial conversion hook actually did nothing with the object because the apiVersion strings include the API group name. Fixed that now and also added test that the returned object are really what we expect. This eseentially matches what we have in e2e tests today. So we might want to strip down the e2e tests to the simplest test which verifies that the plumbing is right with kube-apiserver.

sttts commented:
/assign @liggitt @jpbetz

jpbetz commented:
> Rebased on Joe Betz's test framework.
> 
> Joe Betz please look especially at the 2nd and 3rd last commits. The non-trivial conversion hook actually did nothing with the object because the apiVersion strings include the API group name. Fixed that now and also added test that the returned object are really what we expect.

Ah, I see the problem. Thanks for fixing. The "non-trivial conversion" fixes lgtm. 

 This eseentially matches what we have in e2e tests today. So we might want to strip down the e2e tests to the simplest test which verifies that the plumbing is right with kube-apiserver.

I agree.

liggitt commented:
/retest

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/conversion_test.go:
```diff
@@ -87,8 +85,7 @@ func TestWebhookConverter(t *testing.T) {
 	etcd3watcher.TestOnlySetFatalOnDecodeError(false)
 	defer etcd3watcher.TestOnlySetFatalOnDecodeError(true)
 
-	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, apiextensionsfeatures.CustomResourceWebhookConversion, true)()
```
why this switch? this leaks the feature gate change out of this test

jpbetz replied:
I didn't realize that. How does it get leaked?

sttts replied:
We should really start using server-local feature gates. This is insane.

@jpbetz our feature gates are global. My change of moving it to the flags (which is the right thing to do to avoid custom plumbing), lacks the defer code to set it back to normal.

sttts replied:
fixed.

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/conversion_test.go:
```diff
@@ -406,26 +406,30 @@ func failureResponseConverter(message string) func(review apiextensionsv1beta1.C
-	if currentAPIVersion == "v1beta2" && desiredAPIVersion == "v1beta1" {
+	currentAPIVersion := u.GetAPIVersion()
+
+	fmt.Printf("Converting %s/%s from %s to %s\n", u.GetNamespace(), u.GetName(), currentAPIVersion, desiredAPIVersion)
```
drop debug printing?

sttts replied:
wiring t there is awkward (it changes due to subtests). Will remove it.

liggitt commented:
a couple questions, and could use some squashing

jpbetz reviewed (commented):
Thanks for improving this @sttts. I've learned a couple useful things. I did a full pass on this, didn't see anything to comment on beyond what's already been mentioned.

sttts commented:
> I didn't realize that. How does it get leaked?

@jpbetz we didn't expose the flags in the the server startup. So that feature gate manipulation was the quickest way to get alpha conversion.

sttts commented:
Addressed all comments.

sttts committed:
apiextensions: fix npe in subresource test

sttts committed:
apiextensions: make conversion webhook test server more compositional

sttts committed:
apiextensions: always fix storage version after tests

sttts committed:
apiextensions: port conversion e2e tests to integration

sttts commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/conversion_test.go:
```diff
@@ -159,7 +159,7 @@ func TestWebhookConverter(t *testing.T) {
 				name := fmt.Sprintf("check-%d", i)
 				t.Run(name, func(t *testing.T) {
-					ctc.setAndWaitStorageVersion(t, "v1beta2")
+					defer ctc.setAndWaitStorageVersion(t, "v1beta2")
```
@jpbetz note this ^^: without the defer the storage version is undefined after the check have been run. Hence, the marker wait logic above might not do what we want.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/conversion_test.go:
```diff
@@ -131,7 +133,8 @@ func TestWebhookConverter(t *testing.T) {
 		t.Run(test.group, func(t *testing.T) {
-			tearDown, webhookClientConfig, webhookWaitReady, err := convert.StartConversionWebhookServerWithWaitReady(test.handler)
+			upCh, handler := closeOnCall(test.handler)
+			tearDown, webhookClientConfig, err := convert.StartConversionWebhookServer(handler)
```
@jpbetz I got rid of StartConversionWebhookServerWithWaitReady. Composing this here is simpler and more go-like, using composition.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/convert/webhook.go:
```diff
@@ -25,91 +25,51 @@ import (
-	// config of a custom resource definition does not take effect immediately, this is needed
-	// by the WaitReady returned StartConversionWebhookServerWithWaitReady to detect when a
-	// conversion client config change in the api server has taken effect.
-	path := fmt.Sprintf("/conversionwebhook-%s", uuid.NewUUID())
```
@jpbetz I believe this was never needed because the test server chooses different ports on each run and therefore reuse of the same URL is highly unlikely or even impossible. Do you agree?

liggitt replied:
agree

sttts commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/convert/webhook.go:
```diff
@@ -25,91 +25,51 @@ import (
 	roots := x509.NewCertPool()
 	if !roots.AppendCertsFromPEM(localhostCert) {
-		return nil, nil, fmt.Errorf("Failed to append Cert from PEM")
+		return nil, nil, fmt.Errorf("failed to append Cert from PEM")
```
@jpbetz a nit, little Go-ism: lower-case errors.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/conversion_test.go:
```diff
@@ -530,14 +549,22 @@ var conversionMultiVersionFixture = &apiextensionsv1beta1.CustomResourceDefiniti
 		Scope: apiextensionsv1beta1.NamespaceScoped,
 		Versions: []apiextensionsv1beta1.CustomResourceDefinitionVersion{
 			{
+				// storage version, same schema as v1alpha1
```
@jpbetz note this additional version. It looks unnecessary here, but becomes necessary as soon as we enable validation and pruning. Then the schemas matter and we cannot do noop conversion from v1beta1 to v1beta2, only between v1alpha1 and v1beta1. I pulled this change from https://github.com/kubernetes/kubernetes/pull/77922 here to avoid rewriting all of this again.

sttts commented:
/retest

etcd disappearing during integration, odd.

sttts committed:
apiextensions: move conversion tests into own package

sttts committed:
Update generated files

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/conversion/conversion_test.go:
```diff
@@ -186,98 +196,153 @@ func validateStorageVersion(t *testing.T, ctc *conversionTestContext) {
-			t.Fatal(err)
+		// Ensure copies of an object have the same fields and values at each custom resource definition version regardless of storage version
+		for clientVersion, client := range clients {
+			t.Run(clientVersion, func(t *testing.T) {
```
naming the subtests `mixed-<version>` or `served-<version>`, etc, would make it possible to tell which test failed by the name. would suggest tweaking this in a follow-up, this PR looks reasonable to merge as is and would like to unblock.

sttts replied:
yes, was near to change that as well.

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/conversion/conversion_test.go:
```diff
@@ -186,98 +196,153 @@ func validateStorageVersion(t *testing.T, ctc *conversionTestContext) {
+	ns := ctc.namespace
+
+	for _, createVersion := range ctc.crd.Spec.Versions {
+		t.Run(fmt.Sprintf("getting objects created as %s", createVersion.Name), func(t *testing.T) {
```
this gets used as the test name and should be more succinct (for use with -run, etc). suggest `nontrivial-converted-<version>`. (include in a follow-up PR)

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/conversion/conversion_test.go:
```diff
@@ -186,98 +196,153 @@ func validateStorageVersion(t *testing.T, ctc *conversionTestContext) {
-			client := ctc.versionedClient(ns, version)
-			obj, err := client.Create(newConversionMultiVersionFixture(ns, name, version), metav1.CreateOptions{})
+	for _, listVersion := range ctc.crd.Spec.Versions {
+		t.Run(fmt.Sprintf("listing objects as %s", listVersion.Name), func(t *testing.T) {
```
suggest `nontrivial-converted-list-<version>`. (include in a follow-up PR)

liggitt commented:
/lgtm
/approve

sttts commented:
/retest

`Quota 'SSD_TOTAL_GB' exceeded. Limit: 1500.0 in region us-east1.`

sttts commented:
/hold

https://github.com/kubernetes/kubernetes/pull/77922 includes this. No need for both in the queue.

roycaihw commented:
/area custom-resources
